examples/fluids/problems/newtonian.c

```diff
@@ -309,8 +309,6 @@ PetscErrorCode NS_NEWTONIAN_IG(ProblemData *problem, DM dm, void *ctx, SimpleBC
   CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_vol_ijacobian.qfunction_context);
   CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_inflow.qfunction_context);
   CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_inflow_jacobian.qfunction_context);
-  CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_outflow.qfunction_context);
-  CeedQFunctionContextReferenceCopy(newtonian_ig_context, &problem->apply_outflow_jacobian.qfunction_context);
```

Comment:
I don't know why these weren't dropped while rebasing (They were removed in [this commit](https://github.com/CEED/libCEED/commit/8a94a473032dc6ed59a2cf0afe1d886fbdb591f4)) but the problem is fixed now. @jedbrown , is this good to merge or do we need a test? This post-processing functionality is not supposed to be working with regression tests. I don't know what kind of test would be useful here.

Reply:
What can I run to see it converging? Can it display errors in primitive variables when the simulation is using conservative variables? (Or is that out of scope? I forget much of the discussion.)

Reply:
You can run it with [these arguments](https://github.com/CEED/libCEED/pull/1109#issuecomment-1373950726). Yes, it only displays the errors in primitive variables when the simulation is using conservative variable but not the other way around. That said, the converted errors are too large. This is probably not the best way to compute the errors.

Reply:
I don't understand what the errors mean for dynamic STG, but if I use `blasius_test.yaml`, it diverges for primitive and blows up for conservative. What I'm looking for is one problem where we can grid converge. I'm okay with that not being tested through CI, but we should have one problem where we can see grid convergence for the two formulations.

Reply:
When you say "it blows up", do you mean the convergence test or a single run? Also, I am understanding that grid convergence is the convergence study with mesh refinement, is that right? I remember our convergence study for the final presentation of the class failed as well. I think you suggested that the BC was overshadowing the results and needed a special treatment. Here is the screenshot from our slides:
![Screenshot from 2023-03-16 20-55-46](https://user-images.githubusercontent.com/49916147/225802258-83f13884-dcc8-478b-a9bc-3e9d16cace79.png)

Reply:
```
build/fluids-navierstokes -options_file examples/fluids/tests-output/blasius_test.yaml -state_var conservative

-- Navier-Stokes solver - libCEED + PETSc --
  MPI:
    Host Name                          : kichatna
    Total ranks                        : 1
  Problem:
    Problem Name                       : blasius
    Stabilization                      : none
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
  PETSc:
    Box Faces                          : 3,20,1
    DM MatType                         : aij
    DM VecType                         : standard
    Time Stepping Scheme               : implicit
  Mesh:
    Number of 1D Basis Nodes (P)       : 2
    Number of 1D Quadrature Points (Q) : 2
    Global DoFs                        : 656
    Owned DoFs                         : 656
    DoFs per node                      : 5
    Global nodes (DoFs / 5)            : 131
    Local nodes                        : 168
Time taken for solution (sec): 0.0764067

Relative Error (absolute error norm, true solution norm):
  Conservative variables:
        Component 0: 0.00211302 (0.00258197, 1.22193)
        Component 1: 0.00394782 (0.192959, 48.8773)
        Component 2: 8.16421 (0.130091, 0.0159343)
        Component 3: -nan. (0., 0.)
        Component 4: 0.000187194 (47.4166, 253302.)
  Primitive variables:
        Component 0: 1.00011 (404047., 404002.)
        Component 1: 396.28 (61679.9, 155.647)
        Component 2: 172899. (8319.15, 0.0481156)
        Component 3: -nan. (0., 0.)
        Component 4: 1103.25 (1.75911e+06, 1594.48)
Time integrator CONVERGED_ITS on time step 5 with final time 1e-05
```
So density is nearly constant, but pressure is wrong by order 1 and the x and y velocity are off by orders of magnitude, along with temperature? I think there's still a bug.

Reply:
I think it's a problem that both fields use `elem_restr_q`.
```c
    CeedOperatorSetField(ceed_data->op_to_prim, "q_cons", ceed_data->elem_restr_q, ceed_data->basis_q, CEED_VECTOR_ACTIVE);
    CeedOperatorSetField(ceed_data->op_to_prim, "q_prim", ceed_data->elem_restr_q, CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
```
The `q_cons` input field is evaluated to quadrature points, but then the `q_prim` output is interpreted as though it was at nodes (and also not corrected for multiplicity). Since you're working with local vectors, you could do a naive CPU traversal running the conversion all at nodes or you could apply a nodal map like James is doing for SGS or you could do a (lumped mass?) projection of the primitive variables into an FE space.

I see `ConvertStateVar_NS` doesn't account for multiplicity in the LocalToGlobal.

Reply:
Sorry I've been slow on getting back to this. I think I have fixed multiplicity and dropped `CEED_BASIS_COLLOCATED` above. But the errors haven't changed much.
```
$ build/fluids-navierstokes -options_file examples/fluids/tests-output/blasius_test.yaml -state_var conservative

...

Relative Error (absolute error norm, true solution norm):
  Conservative variables:
        Component 0: 0.00211302 (0.00258197, 1.22193)
        Component 1: 0.00394782 (0.192959, 48.8773)
        Component 2: 8.16421 (0.130091, 0.0159343)
        Component 3: -nan. (0., 0.)
        Component 4: 0.000187194 (47.4166, 253302.)
  Primitive variables:
        Component 0: 1.00033 (101034., 101001.)
        Component 1: 485.939 (19295., 39.7065)
        Component 2: 143100. (1824.88, 0.0127525)
        Component 3: -nan. (0., 0.)
        Component 4: 1371.4 (547956., 399.558)
Time integrator CONVERGED_ITS on time step 5 with final time 1e-05
```
Either I am still missing something in `setuplibceed.c`, or this method is basically wrong (I am trying to convert variables at quadrature points and map them back to nodal space and compute the errors).

Reply:
I don't see `ComputeL2Projection()` being used in anywhere in the code but can I use that function to try projection instead?

Reply:
You recall the Blasius case. If you converted conservative to primitive at quadrature points, you would see the "scallops", which would give a bias when projected into a nodal space. However, I think something is wrong with the projection because that doesn't explain the many orders of magnitude difference reported here.

Reply:
@jedbrown , I tried to explain the problem in this flowchart. Do you think the second proposal would work?
![libCEED_1109 drawio(3)](https://github.com/CEED/libCEED/assets/49916147/a354416b-70d4-496a-8163-7a808341d84c)

Reply:
The quadrature loop in the "current approach" needs a quadrature weight if it's going to be the projection that the math to the right indicates. Sounds like we're just measuring the residual without quadrature weighting (a bug).

The new proposal looks good, and saves the complexity/cost of a consistent projection. The scalloping bias for a nodally-exact field will be present in anything because it's real.

In your last expression, you have 
```
q_source_true = ics.qfunction(final_time)    # this is a `StatePrimitive` per the box above
dQ_source = q_source_soln - q_source_true # can't subtract primitive from conservative
dQ_target = StateFromU(dQ_source, ...)
error_target = w * dQ_target^2
```
However, `StateFromU` needs the full state, not a deviation/difference. I would do the comparison arithmetic in primitive variables, like this.
```
q_source_true_prim = ics.qfunction(final_time)
q_source_soln_prim = StateFromU(q_source_soln_cons, ...)
error_target = w * (q_source_soln_prim - q_source_true_prim)^2
```

Reply:
I see that the naming is indeed confusing. In the last box, the simulation is in conservative variables and by "source" I mean "conservative" in that context and before converting the variables, and "target" would mean "primitive" (state variables after the conversion). I was using the same notation as in the existing implementation. Will change them.

Oops, of course conservative and primitive variables are not linearly related. Thanks!